python/ql/src/Security/CWE-327/Ssl.qll

```diff
@@ -0,0 +1,221 @@
+import python
+import semmle.python.ApiGraphs
+import TlsLibraryModel
+
+class SSLContextCreation extends ContextCreation {
+  override CallNode node;
+
+  SSLContextCreation() { this = API::moduleImport("ssl").getMember("SSLContext").getACall() }
+
+  override DataFlow::CfgNode getProtocol() {
+    result.getNode() in [node.getArg(0), node.getArgByName("protocol")]
+  }
+}
+
+class SSLDefaultContextCreation extends ContextCreation {
+  SSLDefaultContextCreation() {
+    this = API::moduleImport("ssl").getMember("create_default_context").getACall()
+  }
+
+  // Allowed insecure versions are "TLSv1" and "TLSv1_1"
+  // see https://docs.python.org/3/library/ssl.html#context-creation
+  override DataFlow::CfgNode getProtocol() { none() }
+}
+
+/** Gets a reference to an `ssl.Context` instance. */
+private DataFlow::LocalSourceNode sslContextInstance(DataFlow::TypeTracker t) {
+  t.start() and
+  result = API::moduleImport("ssl").getMember(["SSLContext", "create_default_context"]).getACall()
+  or
+  exists(DataFlow::TypeTracker t2 | result = sslContextInstance(t2).track(t2, t))
+}
+
+/** Gets a reference to an `ssl.Context` instance. */
+DataFlow::Node sslContextInstance() {
+  sslContextInstance(DataFlow::TypeTracker::end()).flowsTo(result)
+}
+
+class WrapSocketCall extends ConnectionCreation {
+  override CallNode node;
+
+  WrapSocketCall() {
+    exists(DataFlow::AttrRead call | node.getFunction() = call.asCfgNode() |
+      call.getAttributeName() = "wrap_socket" and
+      call.getObject() = sslContextInstance()
+    )
+  }
+
+  override DataFlow::CfgNode getContext() {
+    result.getNode() = node.getFunction().(AttrNode).getObject()
+  }
+}
```

Comment:
I think you can just stay in the realm of `API::Node` and make things easy on yourself :smile:  (this is written without autocomplete, so maybe a typo will sneak in)

```suggestion
/** Gets a reference to an `ssl.Context` instance. */
API::Node sslContextInstance() {
  result = API::moduleImport("ssl").getMember(["SSLContext", "create_default_context"]).getReturn()
}

class WrapSocketCall extends ConnectionCreation, DataFlow::CallCfgNode {
  WrapSocketCall() {
    this = sslContextInstance().getMember("wrap_socket").getACall()
  }

  override DataFlow::CfgNode getContext() {
    result = this.getFunction().(DataFlow::AttrRead).getObject()
  }
}
```

Reply:
Thanks! I tried this, but dismissed it when I evaluated `sslContextInstance` and all the results had locations on actual calls. I somehow completely forgot that API nodes are "closed under local flow", so that calling `getMember("wrap_socket")` will indeed find something.

Reply:
Worked first try! 💪